File: PlaywireObjC/ads/view/native/CustomNativeAdView.m
//
//  CustomNativeAdView.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/06/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "CustomNativeAdView.h"

@interface CustomNativeAdView ()
@property (strong, nonatomic) UILabel *headlineLabel;
@property (strong, nonatomic) UILabel *bodyLabel;
@property (strong, nonatomic) UIImageView *iconView;
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UILabel *starLabel;
@property (strong, nonatomic) UILabel *priceLabel;
@property (strong, nonatomic) UILabel *storeLabel;
@property (strong, nonatomic) UILabel *advertiserLabel;
@property (strong, nonatomic, readwrite) UIButton *button;
@property (strong, nonatomic) PWNativeViewContent *adContent;
@end

@implementation CustomNativeAdView

- (instancetype)initWithAdContent:(PWNativeViewContent *)adContent {
    self = [super initWithFrame:CGRectZero];
    if (self) {
        _adContent = adContent;
        [self setupView];
    }
    return self;
}

- (instancetype)init {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:@"Use initWithAdContent: instead"
                                 userInfo:nil];
}

- (instancetype)initWithFrame:(CGRect)frame {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:@"Use initWithAdContent: instead"
                                 userInfo:nil];
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:@"Use initWithAdContent: instead"
                                 userInfo:nil];
}

- (void)setupView {
    // Create UI elements
    self.headlineLabel = [[UILabel alloc] init];
    self.headlineLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleTitle2];
    self.headlineLabel.adjustsFontSizeToFitWidth = YES;
    
    self.bodyLabel = [[UILabel alloc] init];
    self.bodyLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleTitle2];
    
    self.iconView = [[UIImageView alloc] init];
    self.imageView = [[UIImageView alloc] init];
    
    self.advertiserLabel = [[UILabel alloc] init];
    self.advertiserLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleTitle3];
    
    self.button = [[UIButton alloc] init];
    [self.button setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    [self.button setTitle:@"Click" forState:UIControlStateNormal];
    
    // Add subviews
    [self addSubview:self.iconView];
    [self addSubview:self.headlineLabel];
    [self addSubview:self.advertiserLabel];
    [self addSubview:self.imageView];
    [self addSubview:self.bodyLabel];
    [self addSubview:self.button];
    
    // Setup constraints
    self.iconView.translatesAutoresizingMaskIntoConstraints = NO;
    self.headlineLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.advertiserLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    self.bodyLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.button.translatesAutoresizingMaskIntoConstraints = NO;
    
    [NSLayoutConstraint activateConstraints:@[
        [self.iconView.widthAnchor constraintEqualToConstant:30],
        [self.iconView.heightAnchor constraintEqualToConstant:30],
        [self.iconView.topAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.topAnchor constant:10],
        [self.iconView.leadingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.leadingAnchor constant:10],
        
        [self.headlineLabel.topAnchor constraintEqualToAnchor:self.iconView.topAnchor],
        [self.headlineLabel.leadingAnchor constraintEqualToAnchor:self.iconView.trailingAnchor constant:10],
        [self.headlineLabel.trailingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.trailingAnchor constant:-10],
        
        [self.advertiserLabel.bottomAnchor constraintEqualToAnchor:self.iconView.bottomAnchor],
        [self.advertiserLabel.leadingAnchor constraintEqualToAnchor:self.iconView.trailingAnchor constant:10],
        [self.advertiserLabel.widthAnchor constraintEqualToConstant:50],
        
        [self.imageView.topAnchor constraintEqualToAnchor:self.advertiserLabel.topAnchor],
        [self.imageView.leadingAnchor constraintEqualToAnchor:self.advertiserLabel.trailingAnchor],
        [self.imageView.widthAnchor constraintEqualToConstant:20],
        [self.imageView.heightAnchor constraintEqualToConstant:20],
        
        [self.bodyLabel.topAnchor constraintEqualToAnchor:self.iconView.bottomAnchor constant:20],
        [self.bodyLabel.leadingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.leadingAnchor],
        [self.bodyLabel.trailingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.trailingAnchor],
        
        [self.button.centerXAnchor constraintEqualToAnchor:self.centerXAnchor],
        [self.button.widthAnchor constraintEqualToConstant:200],
        [self.button.heightAnchor constraintEqualToConstant:50]
    ]];
    
    // Configure with ad content
    self.iconView.image = self.adContent.icon;
    self.headlineLabel.text = self.adContent.headline;
    self.advertiserLabel.text = self.adContent.advertiser;
    self.imageView.image = self.adContent.image;
    self.bodyLabel.text = self.adContent.body;
    
    UIView *mediaView = self.adContent.mediaView;
    if (mediaView) {
        mediaView.translatesAutoresizingMaskIntoConstraints = NO;
        [self addSubview:mediaView];
        
        [NSLayoutConstraint activateConstraints:@[
            [mediaView.topAnchor constraintEqualToAnchor:self.bodyLabel.bottomAnchor constant:10],
            [mediaView.leadingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.leadingAnchor],
            [mediaView.trailingAnchor constraintEqualToAnchor:self.safeAreaLayoutGuide.trailingAnchor],
            [mediaView.heightAnchor constraintEqualToConstant:200],
            
            [self.button.topAnchor constraintEqualToAnchor:mediaView.bottomAnchor constant:10]
        ]];
    }
    
    [self.button setTitle:self.adContent.callToAction forState:UIControlStateNormal];
}


@end

File: PlaywireSwiftUI/Ads/Native/CustomNativeAdView.swift
//
//  CustomNativeAdView.swift
//  PlaywireSDKApps
//
//  Created by Inder Dhir on 10/7/25.
//

import Foundation
import GoogleMobileAds
import Playwire
import UIKit

final class CustomNativeAdView: UIView {
    private var headlineLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title2)
        label.adjustsFontSizeToFitWidth = true
        return label
    }()
    
    private var bodyLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title2)
        return label
    }()
    
    private var iconView = UIImageView()
    private var actionButton = UIButton()
    private var imageView = UIImageView()
    
    private var starLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title3)
        return label
    }()
    
    private var priceLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title3)
        return label
    }()
    
    private var storeLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title3)
        return label
    }()
    
    private var advertiserLabel: UILabel = {
        let label = UILabel()
        label.font = .preferredFont(forTextStyle: .title3)
        return label
    }()
    
    private var contentView = UIView()
    
    let button: UIButton = {
        let button = UIButton()
        button.setTitleColor(.black, for: .normal)
        button.setTitle(String(localized: "Click"), for: .normal)
        return button
    }()
    
    let adContent: PWNativeViewContent
    
    init(adContent: PWNativeViewContent) {
        self.adContent = adContent
        
        super.init(frame: .zero)
        
        setupView()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupView() {
        iconView.translatesAutoresizingMaskIntoConstraints = false
        headlineLabel.translatesAutoresizingMaskIntoConstraints = false
        advertiserLabel.translatesAutoresizingMaskIntoConstraints = false
        imageView.translatesAutoresizingMaskIntoConstraints = false
        bodyLabel.translatesAutoresizingMaskIntoConstraints = false
        button.translatesAutoresizingMaskIntoConstraints = false
        
        addSubview(iconView)
        addSubview(headlineLabel)
        addSubview(advertiserLabel)
        addSubview(imageView)
        addSubview(bodyLabel)
        addSubview(button)

        NSLayoutConstraint.activate([
            iconView.widthAnchor.constraint(equalToConstant: 30),
            iconView.heightAnchor.constraint(equalToConstant: 30),
            iconView.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor, constant: 10),
            iconView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 10),
            
            headlineLabel.topAnchor.constraint(equalTo: iconView.topAnchor),
            headlineLabel.leadingAnchor.constraint(equalTo: iconView.trailingAnchor, constant: 10),
            headlineLabel.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor, constant: -10),
            
            advertiserLabel.bottomAnchor.constraint(equalTo: iconView.bottomAnchor),
            advertiserLabel.leadingAnchor.constraint(equalTo: iconView.trailingAnchor, constant: 10),
            advertiserLabel.widthAnchor.constraint(equalToConstant: 50),
            
            imageView.topAnchor.constraint(equalTo: advertiserLabel.topAnchor),
            imageView.leadingAnchor.constraint(equalTo: advertiserLabel.trailingAnchor),
            imageView.widthAnchor.constraint(equalToConstant: 20),
            imageView.heightAnchor.constraint(equalToConstant: 20),
            
            bodyLabel.topAnchor.constraint(equalTo: iconView.bottomAnchor, constant: 20),
            bodyLabel.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor),
            bodyLabel.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor),
            
            button.centerXAnchor.constraint(equalTo: centerXAnchor),
            button.widthAnchor.constraint(equalToConstant: 200),
            button.heightAnchor.constraint(equalToConstant: 50)
        ])
        
        iconView.image = adContent.icon
        headlineLabel.text = adContent.headline
        advertiserLabel.text = adContent.advertiser
        imageView.image = adContent.image
        bodyLabel.text = adContent.body
        
        let mediaView = adContent.mediaView
        if let mediaView {
            mediaView.translatesAutoresizingMaskIntoConstraints = false
            addSubview(mediaView)
            
            NSLayoutConstraint.activate([
                mediaView.topAnchor.constraint(equalTo: bodyLabel.bottomAnchor, constant: 10),
                mediaView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor),
                mediaView.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor),
                mediaView.heightAnchor.constraint(equalToConstant: 200),
                
                button.topAnchor.constraint(equalTo: mediaView.bottomAnchor, constant: 10)
            ])
        }
        
        button.setTitle(adContent.callToAction, for: .normal)
    }
}
